src/main/dialogs/basic/dialog.js

```diff
@@ -0,0 +1,433 @@
+                "{arguments}.0", // offsetX
+                "{arguments}.1"  // offsetY
+            ]
+        },
+        setBounds: {
+            funcName: "gpii.app.dialog.setBounds",
+            args: [
+                "{that}",
+                "{that}.options.config.restrictions",
+                "{arguments}.0", // width
+                "{arguments}.1", // height
+                "{arguments}.2", // offsetX
+                "{arguments}.3"  // offsetY
+            ]
+        },
+        setRestrictedSize: {
+            funcName: "gpii.app.dialog.setRestrictedSize",
+            args: [
+                "{that}",
+                "{that}.options.config.restrictions",
+                "{arguments}.0", // width
+                "{arguments}.1"  // height
+            ]
+        },
+        _show: {
+            funcName: "gpii.app.dialog._show",
+            args: [
+                "{that}",
+                "{arguments}.0" // showInactive
+            ]
+        },
+        _hide: {
+            this: "{that}.dialog",
+            method: "hide"
+        },
+        show: {
+            funcName: "gpii.app.dialog.show",
+            args: ["{that}"]
+        },
+        hide: {
+            changePath: "isShown",
+            value: false
+        },
+        focus: {
+            this: "{that}.dialog",
+            method: "focus"
+        },
+        close: {
+            this: "{that}.dialog",
+            method: "close"
+        }
+    }
+});
+
+
+/**
+ * Builds a file URL inside the application **Working Directory**.
+ * @param {String} prefixPath - Prefix for the file path, e.g. "src/renderer"
+ * @param {String} suffixPath - Suffix for the file path, e.g. "index.html"
+ * @return {String} The generated URL
+ */
+gpii.app.dialog.buildFileUrl = function (prefixPath, suffixPath) {
+
+    var appHomePath = fluid.module.terms()["gpii-app"];
+
+    return fluid.stringTemplate(
+        "file://%homePath/%prefixPath/%suffixPath",
+        {
+            homePath: appHomePath,
+            prefixPath: prefixPath,
+            suffixPath: suffixPath
+        });
+};
+
+/**
+ * Creates a dialog. This is done upfront to avoid the delay from creating a new
+ * dialog every time the displayed content should change.
+ * @param {Component} that - The `gpii.app.dialog` instance
+ * @param {Object} windowOptions - The raw Electron `BrowserWindow` settings
+ * @param {String} url - The URL to be loaded in the `BrowserWindow`
+ * @param {Object} params -  Options that are to be supplied to the render process of
+ * the newly created BrowserWindow
+ * @return {BrowserWindow} The Electron `BrowserWindow` component
+ */
+gpii.app.dialog.makeDialog = function (that, windowOptions, url, params) {
+    var dialog = new BrowserWindow(windowOptions);
+
+    dialog.loadURL(url);
+
+    // Keep record in the window itself for its wrapping grade
+    dialog.grade = that.options.gradeNames[that.options.gradeNames.length - 1];
```

Comment:
This is pretty dangerous and will be unstable to options distributions from other parts of the application, etc. 
However I haven't yet found what use is made of this field in any case, we should discuss.

Reply:
I've found the use of gradeNames for the linkedWindow mechanism, was this intended to be related?

Reply:
Well, this isn’t directly related to the linkedWindows mechanism but it shares the idea. It is used as a way for creating a two-way relation between a dialog and its wrapping component. It looked like a useful feature if we want to get the wrapping component from a BrowserWindow instance. It was meant to use the most specific grade from the list as identifier, but as you proposed, it would probably be best to replace it with a generated unique identifier as generally there might be more than one instance of a single dialog grade.
As an example of its usage you can see in the branch [for the QSS i18n](https://github.com/danailbd/gpii-app/blob/GPII-3169/src/main/dialogs/basic/dialog.js#L315) where this ID is sent from the renderer through the IPC channel to notify its wrapping component about the component’s readiness. We’ll probably be merging this branch tomorrow and I may change it there to make use of a unique identifier. What do you think?

Reply:
Removing this for now. An improved version of this mechanism will come in together with the QSS i18n branch.